From 5ce79e9808a22766b2314d7842830d43f6e28bdb Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Mon, 19 Jul 2021 16:44:12 +0300
Subject: [PATCH] [SW-47213] gaudi: pll: flush uart before performing wa

During the pci pll workaround we are changing the uart clock. If the
uart is in the middle of a character transmission at this moment
corrupted data will be received on the other side.
in order to avoid that, before touching the PLLs we wait until
uart has finished the last transaction.
Since this wait state might take a little while, we don't consider
this operation as "fast" anymore.

Change-Id: I097a2cdb81916a674f40a94c3bd9747dc631f23e
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/app/preboot/preboot.c b/app/preboot/preboot.c
index 322c0d4..2263a78 100644
--- a/app/preboot/preboot.c
+++ b/app/preboot/preboot.c
@@ -55,9 +55,9 @@
 		.setup = module_pci_setup,
 		.check = module_pci_check,
 		.disable = module_pci_disable,
-		.run = noop,
+		.run = module_pci_run,
 		.enable = module_pci_enable,
-		.run_fast = module_pci_run,
+		.run_fast = module_pci_run_fast,
 		.teardown = module_pci_teardown,
 		.run_noreturn = noop,
 	},
diff --git a/arch/gaudi/arm/pll.c b/arch/gaudi/arm/pll.c
index ba95279..b708e98 100644
--- a/arch/gaudi/arm/pll.c
+++ b/arch/gaudi/arm/pll.c
@@ -10,6 +10,7 @@
 #include <psoc_pci_pll_regs.h>
 #include <preboot.h>
 #include <pll.h>
+#include <uart.h>
 
 #include "common.h"
 
@@ -142,6 +143,12 @@
 void hl_dev_pci_pll_wa(void)
 {
 	/*
+	 * make sure uart is not in the middle of transmission before touching
+	 * the plls, otherwise corrupted data is being sent
+	 */
+	preboot_uart_tx_flush();
+
+	/*
 	 * pci pll div0 frequency must be equal or lower than div2 for pci phy
 	 * fw upgrade.
 	 * after phy fw upgrade is complete and pci link is up we move div0
diff --git a/lib/common/cmd_module.h b/lib/common/cmd_module.h
index a76a9ac..f3db716 100644
--- a/lib/common/cmd_module.h
+++ b/lib/common/cmd_module.h
@@ -71,6 +71,7 @@
 enum cmd_check_res module_pci_check(void);
 void module_pci_disable(void);
 void module_pci_run(void);
+void module_pci_run_fast(void);
 void module_pci_enable(void);
 void module_pci_teardown(void);
 void pci_dev_specific_setup(void);
diff --git a/lib/common/uart.c b/lib/common/uart.c
index 22068ec..44d78f5 100644
--- a/lib/common/uart.c
+++ b/lib/common/uart.c
@@ -10,6 +10,7 @@
 
 #include "preboot.h"
 #include "uart.h"
+#include "pll.h"
 
 #define UART_SR_TXFULL			(1 << 0) /* TX Buffer Full */
 #define UART_SR_RXFULL			(1 << 1) /* RX Buffer Full */
@@ -122,6 +123,20 @@
 	return !!(regmap_readl(mmPSOC_UART_0_STATE) & UART_SR_TXFULL);
 }
 
+/* timeout is 5 chars duration in usec*/
+#define TX_FLUSH_TIMEOUT	((5 / HL_UART_BAUDRATE) * USEC_PER_SEC)
+
+void preboot_uart_tx_flush(void)
+{
+	u64 start = timestamp_get_us();
+
+	/* wait for tx transaction to finish */
+	while (preboot_uart_tx_full()) {
+		if (timeout_expired(start, TX_FLUSH_TIMEOUT))
+			break;
+	}
+}
+
 void preboot_putc_raw(const char ch)
 {
 	if (tx_buf.initialized) {
diff --git a/lib/common/uart.h b/lib/common/uart.h
index b08f86d..41698ab 100644
--- a/lib/common/uart.h
+++ b/lib/common/uart.h
@@ -21,6 +21,7 @@
 void preboot_puts_hex(const char *log, unsigned int val, bool pad);
 void preboot_puts_char(const char *str, unsigned char val);
 bool preboot_uart_tx_full(void);
+void preboot_uart_tx_flush(void);
 int preboot_uart_tstc(void);
 int preboot_uart_getc(void);
 void preboot_uart_rbuf_init(void);
diff --git a/lib/preboot/module_pci.c b/lib/preboot/module_pci.c
index 86159bb..7653b25 100644
--- a/lib/preboot/module_pci.c
+++ b/lib/preboot/module_pci.c
@@ -91,7 +91,7 @@
 	pci_sts.smlh_link_reg_val = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
 
 	if (!pci_sts.pci_pll_initialized)
-		return FOUND_FAST_RETURN;
+		return FOUND_SLOW_RETURN;
 
 	if (pci_sts.flr_reg_val == 1 || pci_sts.hot_rst_reg_val == 1)
 		return FOUND_FAST_RETURN;
@@ -127,12 +127,15 @@
 
 void module_pci_run(void)
 {
-	unsigned int rst_src;
-
 	if (!pci_sts.pci_pll_initialized) {
 		hl_dev_pci_pll_wa();
 		pci_sts.pci_pll_initialized = true;
 	}
+}
+
+void module_pci_run_fast(void)
+{
+	unsigned int rst_src;
 
 	if (pci_sts.flr_reg_val == 1) {
 		pci_dev_specific_flr();
